Mathlib/Combinatorics/SimpleGraph/Turan.lean

```diff
@@ -91,4 +110,175 @@ theorem not_cliqueFree_of_isTuranMaximal (hn : r ≤ Fintype.card V) (hG : G.IsT
+    _ = Fintype.card V - (univ.filter fun b ↦ ¬G.Adj s b).card :=
+      eq_tsub_of_add_eq (filter_card_add_filter_neg_card_eq_card _)
+    _ = _ := by
+      congr; ext; rw [mem_filter]
+      convert Finpartition.mem_part_ofSetoid_iff_rel.symm
+      simp [setoid]
+
+/-- The parts of a Turán-maximal graph form an equipartition. -/
+theorem isEquipartition : h.finpartition.IsEquipartition := by
+  set fp := h.finpartition
+  by_contra hn
+  rw [Finpartition.not_isEquipartition] at hn
+  obtain ⟨large, hl, small, hs, ineq⟩ := hn
+  obtain ⟨w, hw⟩ := fp.nonempty_of_mem_parts hl
+  obtain ⟨v, hv⟩ := fp.nonempty_of_mem_parts hs
+  have large_eq := fp.eq_of_mem_parts hl (fp.part_mem (mem_univ w)) hw (fp.mem_part (mem_univ w))
+  have small_eq := fp.eq_of_mem_parts hs (fp.part_mem (mem_univ v)) hv (fp.mem_part (mem_univ v))
+  apply absurd h
+  rw [IsTuranMaximal]; push_neg; intro cf
+  use G.replaceVertex v w, inferInstance, cf.replaceVertex v w
+  have ha : G.Adj v w := by
+    by_contra hn; rw [h.not_adj_iff_part_eq, ← small_eq, ← large_eq] at hn
+    rw [hn] at ineq; omega
+  rw [G.card_edgeFinset_replaceVertex_of_adj ha,
+    degree_eq_card_sub_part_card h, ← small_eq, degree_eq_card_sub_part_card h, ← large_eq]
+  have : large.card ≤ Fintype.card V := by simpa using card_le_card large.subset_univ
+  omega
+
+lemma card_parts_le : h.finpartition.parts.card ≤ r := by
+  by_contra! l
+  obtain ⟨z, -, hz⟩ := h.finpartition.exists_subset_part_bijOn
+  have ncf : ¬G.CliqueFree z.card := by
+    refine IsNClique.not_cliqueFree ⟨fun v hv w hw hn ↦ ?_, rfl⟩
+    contrapose! hn
+    exact hz.injOn hv hw (by rwa [← h.not_adj_iff_part_eq])
+  rw [Finset.card_eq_of_equiv hz.equiv] at ncf
+  exact absurd (h.1.mono (Nat.succ_le_of_lt l)) ncf
+
+/-- There are `min n r` parts in a graph on `n` vertices satisfying `G.IsTuranMaximal r`.
+`min` handles the `n < r` case, when `G` is complete but still `r + 1`-cliquefree
+for having insufficiently many vertices. -/
+theorem card_parts : h.finpartition.parts.card = min (Fintype.card V) r := by
+  set fp := h.finpartition
+  apply le_antisymm (le_min fp.card_parts_le_card h.card_parts_le)
+  by_contra! l
+  rw [lt_min_iff] at l
+  obtain ⟨x, -, y, -, hn, he⟩ :=
+    exists_ne_map_eq_of_card_lt_of_maps_to l.1 fun a _ ↦ fp.part_mem (mem_univ a)
+  apply absurd h
+  rw [IsTuranMaximal]; push_neg; rintro -
+  have cf : G.CliqueFree r := by
+    simp_rw [← cliqueFinset_eq_empty_iff, cliqueFinset, filter_eq_empty_iff, mem_univ,
+      forall_true_left, isNClique_iff, and_comm, not_and, isClique_iff, Set.Pairwise]
+    intro z zc; push_neg; simp_rw [h.not_adj_iff_part_eq]
+    exact exists_ne_map_eq_of_card_lt_of_maps_to (zc.symm ▸ l.2) fun a _ ↦ fp.part_mem (mem_univ a)
+  use G ⊔ edge x y, inferInstance, cf.sup_edge x y
+  convert Nat.lt.base G.edgeFinset.card
+  convert G.card_edgeFinset_sup_edge _ hn
+  rwa [h.not_adj_iff_part_eq]
+
+/-- **Turán's theorem**, forward direction.
```

Comment:
```suggestion
/-- **Turán's theorem**, forward direction.

```
for the short docstring-long docstring style

Reply:
:+1: